clip_retrieval/clip_client.py

```diff
@@ -0,0 +1,120 @@
+        """
+        self.url = url
+        self.indice_name = indice_name
+        self.use_mclip = use_mclip
+        self.aesthetic_score = aesthetic_score
+        self.aesthetic_weight = aesthetic_weight
+        self.modality = modality.value
+        self.num_images = num_images
+        self.num_result_ids = num_result_ids
+
+    def query(
+        self,
+        text: str = None,
+        image: str = None,
+    ) -> List[Dict]:
+        """
+        Given text or image/s, search for other captions/images that are semantically similar.
+
+        Args:
+            text: text to be searched semantically.
+            image: base64 string of image to be searched semantically
+
+        Returns:
+            List of dictionaries containing the results.
+        """
+        if text and image:
+            raise ValueError("Only one of text or image can be provided.")
+        if text:
+            return self.__search_knn_api__(text=text)
+        elif image:
+            if image.startswith("http"):
+                return self.__search_knn_api__(image_url=image)
+            else:
+                assert Path(image).exists(), f"{image} does not exist."
+                return self.__search_knn_api__(image=image)
+        else:
+            raise ValueError("Either text or image must be provided.")
+
+    def __search_knn_api__(
+        self,
+        text: str = None,
+        image: str = None,
+        image_url: str = None,
+    ) -> List:
+        """
+        This function is used to send the request to the knn service.
+        It represents a direct API call and should not be called directly outside the package.
+
+        Args:
+            text: text to be searched semantically.
+            image: base64 string of image to be searched semantically.
+
+        Returns:
+            List of dictionaries containing the results.
+        """
+        if image:
+            # Convert image to base64 string
+            with open(image, "rb") as image_file:
+                encoded_string = base64.b64encode(image_file.read())
+                image = str(encoded_string.decode("utf-8"))
+        return requests.post(
+            self.url,
+            data=json.dumps(
+                {
+                    "text": text,
+                    "image": image,
+                    "image_url": image_url,
+                    "deduplicate": True,
+                    "use_safety_model": True,
+                    "use_violence_detector": True,
+                    "indice_name": self.indice_name,
+                    "use_mclip": self.use_mclip,
+                    "aesthetic_score": self.aesthetic_score,
+                    "aesthetic_weight": self.aesthetic_weight,
+                    "modality": self.modality,
+                    "num_images": self.num_images,
+                    "num_result_ids": self.num_result_ids,
+                }
+            ),
+        ).json()[
+            0 : self.num_images
```

Comment:
Maybe just set the number of results as num_images from the start ?

Reply:
I should probably learn how these work in the first place ha. What's the difference between `num_result_ids` and `num_images`?

Reply:
something to do with pagination?